docs/development/machine-bootstrap.md

```diff
@@ -0,0 +1,19 @@
+
+## Flow of creating and bootstrapping Machines/Nodes
+The gardener-controller-manager will ensure during the creation/reconciliation flow
+- that the machine-controller-manager gets deployed into the Shoot namespace of the Seed.
+- that proper ``<cloudprovider>machineclass`` objects and its corresponding secrets exists in the Shoot namespace of the Seed. Those secrets contain information to bootstrap the node and will be copied to the machines of the related class.
+- that ``machine-deployments`` for the in the Shoot manifest specified worker groups exists. Those deployments will spawn the required ``machine`` objects in the Shoot namespace of the Seed.
+- that a valid bootstrap-token secret exists in the kube-system namespace of the Shoot.
+- that the cloud-config configuration for setting up nodes exists in the kube-system namespace of the Shoot.
+
+The machine-controller-manager will pick up the ``machine`` objects and create the vms on the respective infrastructure provider. The bootstrap of a vm, which should join the Shoot cluster as node, is done via CoreOS [cloud-config](https://coreos.com/os/docs/latest/cloud-config.html).
+
+The cloud-config configuration will be downloaded from the Shoot apiserver. This will be archieved by copying the contents of the machineclass secret onto the machine. This secret contains configuration to register a systemd service ``cloud-config-downloader`` and a kubeconfig to fetch only the cloud-config and the bootstrap-token from the Shoot apiserver. The downloaded cloud-config configuration will register additional systemd services for the kubelet and other required services on the node.
+
+The kubelet requires a suitable kubeconfig to join the cluster and it will generate it by relying on the [tls-bootstraping](https://kubernetes.io/docs/reference/command-line-tools-reference/kubelet-tls-bootstrapping/) functionality of the kubelet.
+
+#### Details
+The ``cloud-config-downloader`` fetches the bootstap-token from Shoot apiserver and checks, if already a kubeconfig file exists on the path, which the kubelet configures via its ``--kubeconfig`` flag. If the file exists the downloader does nothing, because it assumes that a proper kubeconfig already exists. If the file not exists, then a bootstrap-kubeconfig will be created with the bootstrap token to authenticate. This bootstrap-kubeconfig will be placed on the path configured via kubelets [``--bootstrap-kubeconfig``](https://kubernetes.io/docs/reference/command-line-tools-reference/kubelet/) flag. Afterwards, the kubelet will perform the tls-bootstrapping process to generate a suitable kubeconfig, which will be stored on the path, which kubelets ``--kubeconfig`` flag is pointing to.
```

Comment:
s/apiserver/API server

I would also mention that the cloud-config-downloader periodically checks for updates and, if detected, restarts all services. That's how K8s patch updates work for the kubelet.